File: services/tradelog/tradelog-api/src/main/java/com/example/tradelog/api/converter/protobuf/ShareJournalModelConverter.java
package com.example.tradelog.api.converter.protobuf;


import java.util.function.Function;

public class ShareJournalModelConverter {
    public static Function<com.example.tradelog.api.spec.model.ShareJournalModel, com.example.tradelog.api.core.model.ShareJournalModel> toModel = protoModel -> {
        return new com.example.tradelog.api.core.model.ShareJournalModel(
                TransactionModelConverter.toModel.apply(protoModel.getTransactionDetails()),
                protoModel.getPrice(),
                protoModel.getActualPrice(),
                protoModel.getQuantity(),
                ActionTypeConverter.toModel.apply(protoModel.getAction())
        );
    };

    public static Function<com.example.tradelog.api.core.model.ShareJournalModel, com.example.tradelog.api.spec.model.ShareJournalModel> toProto = apiModel -> {
        return com.example.tradelog.api.spec.model.ShareJournalModel.newBuilder()
                .setTransactionDetails(TransactionModelConverter.toProto.apply(apiModel.getTransactionDetails()))
                .setPrice(apiModel.getPrice())
                .setActualPrice(apiModel.getActualPrice())
                .setQuantity(apiModel.getQuantity())
                .setAction(ActionTypeConverter.toProto.apply(apiModel.getAction()))
                .build();
    };
}

File: services/tradelog/tradelog-api/src/main/kotlin/com/example/tradelog/api/rest/converter/ShareJournalModelConverter.kt
package com.example.tradelog.api.rest.converter

import com.example.tradelog.api.core.model.ActionType
import com.example.tradelog.api.core.model.ShareJournalModel
import com.example.tradelog.api.spec.model.TLShareJournalDto

class ShareJournalModelConverter {

    companion object {
        fun toModel(dto: TLShareJournalDto): ShareJournalModel {
            return ShareJournalModel(
                    transactionDetails = TransactionModelConverter.toModel(dto.transactionDetails),
                    price = dto.price,
                    quantity = dto.quantity,
                    action = ActionType.lookup(dto.action.name)
            )
        }

        fun toDto(model: ShareJournalModel): TLShareJournalDto {
            return TLShareJournalDto.newBuilder()
                    .setTransactionDetails(TransactionModelConverter.toDto(model.transactionDetails))
                    .setPrice(model.price)
                    .setQuantity(model.quantity)
                    .setAction(TLShareJournalDto.ActionType.valueOf(model.action.name))
                    .build()
        }
    }
}
